src/crimson/osd/object_context.h

```diff
@@ -137,22 +137,34 @@ class ObjectContext : public ceph::common::intrusive_lru_base<
 private:
   template <typename Lock, typename Func>
   auto _with_lock(Lock& lock, Func&& func) {
-    Ref obc = this;
     auto maybe_fut = lock.lock();
     return seastar::futurize_invoke([
         maybe_fut=std::move(maybe_fut),
-        func=std::forward<Func>(func)]() mutable {
+        func=std::forward<Func>(func),
+	obc=Ref(this),
+	&lock]() mutable {
       if (maybe_fut) {
         return std::move(*maybe_fut
-        ).then([func=std::forward<Func>(func)]() mutable {
-          return seastar::futurize_invoke(func);
+        ).then([func=std::forward<Func>(func), obc, &lock]() mutable {
+          return seastar::futurize_invoke(
+	    func
+	  ).finally([&lock, obc] {
+	    /* We chain the finally block here because it's possible for
+	     * *maybe_fut from lock.lock() above to fail due to a call to
+	     * ObjectContext::interrupt, which calls tri_mutex::abort.
+	     * In the event of such an error, the lock isn't actually taken
+	     * and calling unlock() would be incorrect. */
```

Comment:
A surprising corner case.

Probably it would be easier not to interrupt during locking?

Reply:
> A surprising corner case.
> 
> Probably it would be easier not to interrupt during locking?

Shouldn't we be able to interrupt during locking?

Reply:
Not sure, if interruption is supposed to be very rare, probably some utilities can be simplified without worrying about interruption and less error-prone.

In this case, for example, if delaying the interruption after being locked can still be correct and rare (probably I'm wrong), simplifying might be a choice.

Anyway, the fix should be good already.